File: backend/src/routers/tasks.py
from typing import Annotated
import sqlite3

from fastapi import APIRouter, Depends, HTTPException, status

from ..database import get_db, row_to_dict, utc_now
from ..deps import current_user
from ..pydantic_models import TaskCreateRequest, TaskPublic, TaskUpdateRequest
from ..serializers import serialize_task


router = APIRouter(prefix="/api/tasks", tags=["tasks"])


def _get_task_or_404(db: sqlite3.Connection, task_id: int) -> dict:
    task = row_to_dict(
        db.execute("SELECT * FROM tasks WHERE id = ?", (task_id,)).fetchone()
    )
    if not task:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Задача не найдена",
        )
    return task


def _ensure_can_modify_task(user: dict, task: dict) -> None:
    if user["role"] != "admin" and task["user_id"] != user["id"]:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Нет прав на изменение этой задачи",
        )


@router.get("/", response_model=list[TaskPublic])
def list_tasks(
    user: Annotated[dict, Depends(current_user)],
    db: Annotated[sqlite3.Connection, Depends(get_db)],
) -> list[TaskPublic]:
    if user["role"] == "admin":
        rows = db.execute(
            """
            SELECT t.*, u.name AS owner_name, u.login AS owner_login
            FROM tasks t
            LEFT JOIN users u ON t.user_id = u.id
            ORDER BY t.created_at DESC, t.id DESC
            """
        ).fetchall()
    else:
        rows = db.execute(
            """
            SELECT * FROM tasks
            WHERE user_id = ?
            ORDER BY created_at DESC, id DESC
            """,
            (user["id"],),
        ).fetchall()
    return [serialize_task(dict(row)) for row in rows]


@router.post("/", response_model=TaskPublic, status_code=status.HTTP_201_CREATED)
def create_task(
    payload: TaskCreateRequest,
    user: Annotated[dict, Depends(current_user)],
    db: Annotated[sqlite3.Connection, Depends(get_db)],
) -> TaskPublic:
    now = utc_now()
    cursor = db.execute(
        """
        INSERT INTO tasks (title, done, user_id, created_at, updated_at)
        VALUES (?, 0, ?, ?, ?)
        """,
        (payload.title.strip(), user["id"], now, now),
    )
    task = row_to_dict(
        db.execute("SELECT * FROM tasks WHERE id = ?", (cursor.lastrowid,)).fetchone()
    )
    return serialize_task(task)


@router.patch("/{task_id}", response_model=TaskPublic)
def update_task(
    task_id: int,
    payload: TaskUpdateRequest,
    user: Annotated[dict, Depends(current_user)],
    db: Annotated[sqlite3.Connection, Depends(get_db)],
) -> TaskPublic:
    task = _get_task_or_404(db, task_id)
    _ensure_can_modify_task(user, task)

    db.execute(
        "UPDATE tasks SET done = ?, updated_at = ? WHERE id = ?",
        (1 if payload.done else 0, utc_now(), task_id),
    )
    updated = row_to_dict(
        db.execute("SELECT * FROM tasks WHERE id = ?", (task_id,)).fetchone()
    )
    return serialize_task(updated)


@router.delete("/{task_id}")
def delete_task(
    task_id: int,
    user: Annotated[dict, Depends(current_user)],
    db: Annotated[sqlite3.Connection, Depends(get_db)],
) -> dict:
    task = _get_task_or_404(db, task_id)
    _ensure_can_modify_task(user, task)
    db.execute("DELETE FROM tasks WHERE id = ?", (task_id,))
    return {"message": "Задача удалена"}

File: backend-node/routes/tasks.js
const express = require('express');
const router = express.Router();
const db = require('../db');
const { requireAuth, requireRole } = require('../middleware/auth');

// READ — получить задачи
// Админ видит все задачи, пользователь — только свои
router.get('/', requireAuth, async (req, res) => {
  try {
    let query;
    let params;

    if (req.user.role === 'admin') {
      // Админ видит все задачи со информацией о владельце
      query = `
        SELECT t.*, u.name as owner_name, u.login as owner_login 
        FROM tasks t 
        LEFT JOIN users u ON t.user_id = u.id 
        ORDER BY t.created_at DESC
      `;
      params = [];
    } else {
      // Пользователь видит только свои задачи
      query = 'SELECT * FROM tasks WHERE user_id = $1 ORDER BY created_at DESC';
      params = [req.user.id];
    }

    const result = await db.query(query, params);
    res.json(result.rows);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

// CREATE — создать задачу
router.post('/', requireAuth, async (req, res) => {
  const { title } = req.body;
  if (!title) return res.status(400).json({ message: 'Нужно название задачи' });
  
  try {
    const result = await db.query(
      'INSERT INTO tasks (title, user_id) VALUES ($1, $2) RETURNING *',
      [title, req.user.id]
    );
    res.status(201).json(result.rows[0]);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

// UPDATE — изменить задачу
// Пользователь может изменять только свои задачи, админ — любые
router.patch('/:id', requireAuth, async (req, res) => {
  const { id } = req.params;
  const { done } = req.body;
  
  try {
    // Проверяем права
    const checkResult = await db.query('SELECT user_id FROM tasks WHERE id = $1', [id]);
    
    if (checkResult.rows.length === 0) {
      return res.status(404).json({ message: 'Задача не найдена' });
    }

    const taskOwnerId = checkResult.rows[0].user_id;
    
    // Если не админ и не владелец — запрещаем
    if (req.user.role !== 'admin' && req.user.id !== taskOwnerId) {
      return res.status(403).json({ message: 'Нет прав на редактирование этой задачи' });
    }

    const result = await db.query(
      'UPDATE tasks SET done = $1 WHERE id = $2 RETURNING *',
      [done, id]
    );

    res.json(result.rows[0]);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

// DELETE — удалить задачу
// Пользователь может удалять только свои задачи, админ — любые
router.delete('/:id', requireAuth, async (req, res) => {
  const { id } = req.params;
  
  try {
    // Проверяем права
    const checkResult = await db.query('SELECT user_id FROM tasks WHERE id = $1', [id]);
    
    if (checkResult.rows.length === 0) {
      return res.status(404).json({ message: 'Задача не найдена' });
    }

    const taskOwnerId = checkResult.rows[0].user_id;
    
    // Если не админ и не владелец — запрещаем
    if (req.user.role !== 'admin' && req.user.id !== taskOwnerId) {
      return res.status(403).json({ message: 'Нет прав на удаление этой задачи' });
    }

    await db.query('DELETE FROM tasks WHERE id = $1', [id]);
    res.json({ message: 'Задача удалена' });
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

module.exports = router;